lib/SILGen/SILGenType.cpp

```diff
@@ -619,13 +615,12 @@ class SILGenConformance : public SILGenWitnessTable<SILGenConformance> {
     return Conformance->getWitness(decl);
   }
 
-  void addPlaceholder(MissingMemberDecl *placeholder) {
-    llvm_unreachable("generating a witness table with placeholders in it");
-  }
-
-  void addMissingMethod(SILDeclRef requirement) {
-    llvm_unreachable("generating a witness table with placeholders in it");
-  }
+  // Treat placeholders and missing methods as no-ops. These may be encountered
+  // during lazy typechecking when SILGen triggers witness resolution and
+  // discovers and invalid conformance. The diagnostics emitted during witness
+  // resolution should cause compilation to fail.
+  void addPlaceholder(MissingMemberDecl *placeholder) {}
+  void addMissingMethod(SILDeclRef requirement) {}
```

Comment:
Can we add asserts about that lazy type checking is enabled in these function bodies?

Reply:
I think such an assert would be a liability in the future as we move towards more laziness by default in the compiler. In my opinion we should not assert unless we can guarantee that there is a fatal problem. We often have trouble building real-world projects with asserts compilers because of ambitious asserts, and this feels to me like it would be such an assert.

Reply:
> this feels to me like it would be such an assert.
IIUC, are you saying that we may already step on the `llvm_unreachable` case nowadays?

Reply:
No, I'm saying that I think leaving an assert in here predicated on a flag would be too fragile to be worth it. This is not the appropriate stage of compilation to decide whether or not this condition is unexpected and crash the compiler. The condition that ought to be checked is "did we fail to emit this witness and also not emit a diagnostic?". That would be better done at a later stage, where we can assume that if we've made it that far we didn't emit any fatal diagnostics during SILGen.

Reply:
OK. I'm just a little bit concerning about the potential loss of the qualification and documentation capability of an `unreachable` statement without providing alternatives back in place.